sklearn/metrics/tests/test_classification.py

```diff
@@ -3095,3 +3102,26 @@ def test_d2_log_loss_score_raises():
     err = "The labels array needs to contain at least two"
     with pytest.raises(ValueError, match=err):
         d2_log_loss_score(y_true, y_pred, labels=labels)
+
+
+@pytest.mark.parametrize(
+    "array_namespace, device, _", yield_namespace_device_dtype_combinations()
+)
+def test_confusion_matrix_array_api(array_namespace, device, _):
+    """Test that `confusion_matrix` works for all array types if need_index_conversion
+    evaluates to `True`and that it raises if not at least one label from `y_pred` is in
+    `y_true`."""
+    xp = _array_api_for_tests(array_namespace, device)
+
+    y_true = xp.asarray([1, 2, 3], device=device)
+    y_pred = xp.asarray([4, 5, 6], device=device)
+
+    with config_context(array_api_dispatch=True):
+        result = confusion_matrix(y_true, y_pred)
+        xp_result, _ = get_namespace(result)
+        assert _is_numpy_namespace(xp_result)
+
+        # Since the computation always happens with NumPy / SciPy on the CPU, this
+        # function is expected to return an array allocated on the CPU even when it does
+        # not match the input array's device.
+        assert result.device == "cpu"
```

Comment:
I have adjusted this test to your suggestions from [this comment](https://github.com/scikit-learn/scikit-learn/pull/30440#discussion_r1897921072), @ogrisel. But here,  the test is narrower because we made the return value of `confusion_matrix` to always be a numpy array on cpu.